### PR TITLE
Move DeviceInfoProvider to common code and use it in linux and NRF.

### LINKED ISSUE #17596: [SQA]-THREAD In (TC-LO-2.1) Supported-locals has no entries.
**Problem :**
While executing the TC-LO-2.1 in Thread platform **Supported-locals** has no entries, But while executing in the RPI platform it throws the output as 8 entries.

**Expected behavior:**   SupportedLocales: 8 entries
[1650539057.043007][2885:2890] CHIP:TOO:     [1]: en-US
[1650539057.043038][2885:2890] CHIP:TOO:     [2]: de-DE
[1650539057.043067][2885:2890] CHIP:TOO:     [3]: fr-FR
[1650539057.043095][2885:2890] CHIP:TOO:     [4]: en-GB
[1650539057.043124][2885:2890] CHIP:TOO:     [5]: es-ES
[1650539057.043152][2885:2890] CHIP:TOO:     [6]: zh-CN
[1650539057.043181][2885:2890] CHIP:TOO:     [7]: it-IT
[1650539057.043209][2885:2890] CHIP:TOO:     [8]: ja-JP

**Actual behavior:**  SupportedLocales: 0 entries

**Steps to reproduce:**  `./chip-tool localizationconfiguration read supported-locales 33 0`

**Platform:** Thread

**CommitID** used:  20e1a4a705770d5222a8ebf04e12ab2ce0304df6

**Reference:**  [Spec](https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/service_device_management/LocalizationConfiguration.adoc#supportedlocales-attribute) , [Test-plan](https://github.com/CHIP-Specifications/chip-test-plans/blob/master/src/cluster/Localization.adoc#tc-lo-2-1localization-configuration-cluster-attributes)

**Logs:**

**In RPI:**

> [1650539057.042853][2885:2890] CHIP:TOO: Endpoint: 0 Cluster: 0x0000_002B Attribute 0x0000_0002 DataVersion: 2961673245
[1650539057.042945][2885:2890] CHIP:TOO:   SupportedLocales: 8 entries
[1650539057.043007][2885:2890] CHIP:TOO:     [1]: en-US
[1650539057.043038][2885:2890] CHIP:TOO:     [2]: de-DE
[1650539057.043067][2885:2890] CHIP:TOO:     [3]: fr-FR
[1650539057.043095][2885:2890] CHIP:TOO:     [4]: en-GB
[1650539057.043124][2885:2890] CHIP:TOO:     [5]: es-ES
[1650539057.043152][2885:2890] CHIP:TOO:     [6]: zh-CN
[1650539057.043181][2885:2890] CHIP:TOO:     [7]: it-IT
[1650539057.043209][2885:2890] CHIP:TOO:     [8]: ja-JP
[1650539057.043330][2885:2890] CHIP:EM: Sending Standalone Ack for MessageCounter:11868231 on exchange 15546i

**In Thread:**

> [1650538116.572188][29974:29979] CHIP:EM: Removed CHIP MessageCounter:9183892 from RetransTable on exchange 1536i
[1650538116.572312][29974:29979] CHIP:DMG: ReportDataMessage =
[1650538116.572377][29974:29979] CHIP:DMG: {
[1650538116.572431][29974:29979] CHIP:DMG: 	AttributeReportIBs =
[1650538116.572513][29974:29979] CHIP:DMG: 	[
[1650538116.572574][29974:29979] CHIP:DMG: 		AttributeReportIB =
[1650538116.572657][29974:29979] CHIP:DMG: 		{
[1650538116.572726][29974:29979] CHIP:DMG: 			AttributeDataIB =
[1650538116.572789][29974:29979] CHIP:DMG: 			{
[1650538116.572875][29974:29979] CHIP:DMG: 				DataVersion = 0x7a5e82db,
[1650538116.572947][29974:29979] CHIP:DMG: 				AttributePathIB =
[1650538116.573020][29974:29979] CHIP:DMG: 				{
[1650538116.573108][29974:29979] CHIP:DMG: 					Endpoint = 0x0,
[1650538116.573195][29974:29979] CHIP:DMG: 					Cluster = 0x2b,
[1650538116.573273][29974:29979] CHIP:DMG: 					Attribute = 0x0000_0002,
[1650538116.573354][29974:29979] CHIP:DMG: 				}
[1650538116.573432][29974:29979] CHIP:DMG: 					
[1650538116.573512][29974:29979] CHIP:DMG: 					Data = [
[1650538116.573595][29974:29979] CHIP:DMG: 						
[1650538116.573687][29974:29979] CHIP:DMG: 					],
[1650538116.573767][29974:29979] CHIP:DMG: 			},
[1650538116.573847][29974:29979] CHIP:DMG: 			
[1650538116.574039][29974:29979] CHIP:DMG: 		},
[1650538116.574125][29974:29979] CHIP:DMG: 		
[1650538116.574184][29974:29979] CHIP:DMG: 	],
[1650538116.574258][29974:29979] CHIP:DMG: 	
[1650538116.574318][29974:29979] CHIP:DMG: 	SuppressResponse = true, 
[1650538116.574381][29974:29979] CHIP:DMG: 	InteractionModelRevision = 1
[1650538116.574452][29974:29979] CHIP:DMG: }
[1650538116.574803][29974:29979] CHIP:TOO: Endpoint: 0 Cluster: 0x0000_002B Attribute 0x0000_0002 DataVersion: 2053014235
[1650538116.574980][29974:29979] CHIP:TOO:   SupportedLocales: 0 entries
[1650538116.575210][29974:29979] CHIP:EM: Sending Standalone Ack for MessageCounter:10443902 on exchange 1536i

### PR DESCRIPTION
**Pull request by andy31415**
#### Problem
#17596 reports that DeviceInfoProvider is not supported in nrfconnect (specifically supported locales list is empty).

This fixes #17596.

#### Change overview
Move DeviceInfoProvider implementation as an example provider and make use of it in linux and nrf.

Tried to preserve existing behavior and have linux (and darwin app main as a sideffect) support deviceinfo provider. Ameba uses a custom build script that is out of the tree, so I could not update that one).

#### Testing
Ran

```sh
#!/usr/bin/env bash

set -ex

function finish {
  echo "STATUS: $?"
}

trap finish EXIT

export NODEID=$RANDOM

export THREAD_CRED=hex:... # my test thread credentials here
export CHIP_TOOL="./out/linux-x64-chip-tool-ipv6only/chip-tool"

log_file="run_log_$(date +'%Y-%m-%d_%H-%M-%S').txt"
rm /tmp/chip_*

echo "NODE ID USED: $NODEID" | tee $log_file

echo "=============================================================" >>$log_file
echo $CHIP_TOOL pairing ble-thread $NODEID $THREAD_CRED 20202021 3840 | tee -a $log_file
$CHIP_TOOL pairing ble-thread $NODEID $THREAD_CRED 20202021 3840  >>$log_file

echo "=============================================================" >>$log_file
echo $CHIP_TOOL localizationconfiguration read supported-locales $NODEID 0 | tee -a $log_file
$CHIP_TOOL localizationconfiguration read supported-locales $NODEID 0 | tee -a $log_file | grep 'CHIP:DMG:'
```

and saw output that was non-empty:

```
...
[1653067684.692432][838470:838477] CHIP:DMG:                    AttributeDataIB =
[1653067684.692434][838470:838477] CHIP:DMG:                    {
[1653067684.692436][838470:838477] CHIP:DMG:                            DataVersion = 0x2179c2af,
[1653067684.692438][838470:838477] CHIP:DMG:                            AttributePathIB =
[1653067684.692440][838470:838477] CHIP:DMG:                            {
[1653067684.692442][838470:838477] CHIP:DMG:                                    Endpoint = 0x0,
[1653067684.692444][838470:838477] CHIP:DMG:                                    Cluster = 0x2b,
[1653067684.692447][838470:838477] CHIP:DMG:                                    Attribute = 0x0000_0001,
[1653067684.692448][838470:838477] CHIP:DMG:                                    ListIndex = Null,
[1653067684.692450][838470:838477] CHIP:DMG:                            }
[1653067684.692453][838470:838477] CHIP:DMG:
[1653067684.692456][838470:838477] CHIP:DMG:                                    Data = "ja-JP",
[1653067684.692458][838470:838477] CHIP:DMG:                    },
[1653067684.692462][838470:838477] CHIP:DMG:
[1653067684.692463][838470:838477] CHIP:DMG:            },
[1653067684.692468][838470:838477] CHIP:DMG:
[1653067684.692470][838470:838477] CHIP:DMG:    ],
[1653067684.692494][838470:838477] CHIP:DMG:
[1653067684.692497][838470:838477] CHIP:DMG:    SuppressResponse = true,
[1653067684.692499][838470:838477] CHIP:DMG:    InteractionModelRevision = 1
[1653067684.692501][838470:838477] CHIP:DMG: }
[1653067684.693219][838470:838470] CHIP:DMG: IM WH moving to [Uninitialized]
[1653067684.693221][838470:838470] CHIP:DMG: IM WH moving to [Uninitialized]
[1653067684.693222][838470:838470] CHIP:DMG: IM WH moving to [Uninitialized]
[1653067684.693223][838470:838470] CHIP:DMG: IM WH moving to [Uninitialized]
[1653067684.693225][838470:838470] CHIP:DMG: All ReadHandler-s are clean, clear GlobalDirtySet
+ finish
+ echo 'STATUS: 0'
STATUS: 0
```

### CODE REVIEW

File: src/platform/Zephyr/DeviceInfoProviderImpl.cpp [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+}
+
+DeviceInfoProviderImpl::FixedLabelIteratorImpl::FixedLabelIteratorImpl(EndpointId endpoint) : mEndpoint(endpoint)
+{
+    mIndex = 0;
+}
+
+size_t DeviceInfoProviderImpl::FixedLabelIteratorImpl::Count()
+{
+    // In Zephyr Simulation, return the size of the hardcoded labelList on all endpoints.
```
Comment (bzbarsky-apple) [inline src/platform/Zephyr/DeviceInfoProviderImpl.cpp:56]: What is "Zephyr Simulation"?  How would an actual device maker implement an actual fixed label cluster on top of this code?

File: src/platform/Zephyr/DeviceInfoProviderImpl.cpp [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+
+    return mStorage->SyncSetKeyValue(keyAlloc.UserLabelLengthKey(endpoint), &val, static_cast<uint16_t>(sizeof(val)));
+}
+
+CHIP_ERROR DeviceInfoProviderImpl::GetUserLabelLength(EndpointId endpoint, size_t & val)
+{
+    DefaultStorageKeyAllocator keyAlloc;
+    uint16_t len = static_cast<uint16_t>(sizeof(val));
+
+    return mStorage->SyncGetKeyValue(keyAlloc.UserLabelLengthKey(endpoint), &val, len);
```
Comment (bzbarsky-apple) [inline src/platform/Zephyr/DeviceInfoProviderImpl.cpp:132]: This allows unbounded storage.

I think all the comments at https://github.com/project-chip/connectedhomeip/pull/18236#pullrequestreview-968677986 apply to this copy/paste.....

**Review by bzbarsky-apple — DISMISSED** [review]

File: src/platform/Zephyr/DeviceInfoProviderImpl.cpp (hunk 1 above)
Comment (andy31415) [reply src/platform/Zephyr/DeviceInfoProviderImpl.cpp:56]: It is basically a copy & paste from other platforms - I see amebad and linux being basically the same thing.

As I understand it, each application needs to implement this, however the paths to bind this at application level seems to not be implemented and this exists rather at the platform level.

If we are to re-make the bindings, I believe this should happen as a separate PR/effort. Right now this is "do the same as linux/darwin/ameba".

Comment (bzbarsky-apple) [reply src/platform/Zephyr/DeviceInfoProviderImpl.cpp:56]: I know it's a copy/paste.  My point is that for Linux and Darwin this approach maybe made sense in the "this is just test code, no one is going to actually use it" sense, but in general this approach does not make sense, and checking it in just gives us a false sense of "it's implemented" when it's actually not.

Let me turn the question around: what is the point of adding this code?  As far as I can tell, the only point is to make it looks like tests pass when they would not actually do that for a real application.  I don't think we should be going there.

Comment (andy31415) [reply src/platform/Zephyr/DeviceInfoProviderImpl.cpp:56]: Chatted offline, I will move the existing implementation (which seems to be copied and pasted across implementations) to a central place that can be reused by various app main.

**Review by LuDuda — DISMISSED** [review]
Thank you @andy31415 for contribution.

I agree with Boris, that this should not be implemented in the platform layer.

Providing example implementation of DeviceInfoProvider for all platforms seems to be better solution (similarly what we do in attestation credentials `SetDeviceAttestationCredentialsProvider(Examples::GetExampleDACProvider());`)

File: src/platform/Zephyr/DeviceInfoProviderImpl.cpp (hunk 1 above)
Comment (andy31415) [reply src/platform/Zephyr/DeviceInfoProviderImpl.cpp]: Updated to a common implementation.

**Comment by andy31415** [comment]
> Thank you @andy31415 for contribution.
> 
> I agree with Boris, that this should not be implemented in the platform layer.
> 
> Providing example implementation of DeviceInfoProvider for all platforms seems to be better solution (similarly what we do in attestation credentials `SetDeviceAttestationCredentialsProvider(Examples::GetExampleDACProvider());`)

@LuDuda I just updated the PR and moved things to a common layer.

NRF uses a custom cmake build which required a bit of updating. Please take a look.

File: examples/providers/BUILD.gn [hunk 3]
```diff
@@ -0,0 +1,37 @@
+# Copyright (c) 2021 Project CHIP Authors
```
Comment (Damian-Nordic) [inline examples/providers/BUILD.gn:1]: 2022

**Review by Damian-Nordic — APPROVED** [review]

**Review by LuDuda — APPROVED** [review]

**Review by tcarmelveilleux — APPROVED** [review]

**Pull request merged**